File: leetcode/april2024/c++/11_remove_k_digits.cpp
// CODE

class Solution {
    public:
    string removeKdigits(string num, int k) {
        string res;
        for(char c: num) {
            // keeping only n - k smaller elements in stack i.e., res string
            while((res.size()) and (c < res.back()) and k) res.pop_back(), k--;
            if(res.size() + (c - '0')) res.push_back(c);
        }
        while(k-- and res.size()) res.pop_back();
        return res.size() ? res : "0";
    }
};

File: leetcode/april2024/python/11_remove_k_digits.py
# CODE

class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        res = []
        for c in num:
            # keeping only n - k smaller elements in stack i.e., res list
            while res and c < res[-1] and k:
                res.pop()
                k -= 1
            if res or c != '0':
                res.append(c)
        while k and res:
            res.pop()
            k -= 1
        return '0' if not res else ''.join(res)